README
gst-plugin-local
----------------

This Plugins aims to be a replacement for the inter* Elements that better handles
A/V Synchronisation and unreliable Sources.

It's made for [Voctomix](https://github.com/voc/voctomix/) but should be of use
for any Gstreamer-System that processes live Video with unreliable Sources.


ChangeLog
2016-08-06  Peter Körner <peter@mazdermind.de>

    * continue work

2015-12-25  Peter Körner <peter@mazdermind.de>

    * buildable (but still totally broken) version of an interconnect element
    * Whitepaper of functionality


src/gstlocal.c
/* GStreamer
 * Copyright (C) 2015 Peter Körner <peter@mazdermind.de>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Suite 500,
 * Boston, MA 02110-1335, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "gstlocalsink.h"

static gboolean
plugin_init (GstPlugin * plugin)
{
  gst_element_register (plugin, "localsink", GST_RANK_NONE,
      GST_TYPE_LOCAL_SINK);

  return TRUE;
}

GST_PLUGIN_DEFINE (GST_VERSION_MAJOR,
    GST_VERSION_MINOR,
    local,
    "plugin for timestamp aware inter-pipeline communication",
    plugin_init, VERSION, "LGPL", PACKAGE_NAME, "https://github.com/MaZderMind/gst-plugin-local")


src/gstlocalsink.h
/* GStreamer
 * Copyright (C) 2011 David A. Schleef <ds@schleef.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifndef _GST_LOCAL_SINK_H_
#define _GST_LOCAL_SINK_H_

#include <gst/base/gstbasesink.h>
#include "gstlocalsurface.h"

G_BEGIN_DECLS
#define GST_TYPE_LOCAL_SINK           (gst_local_sink_get_type())
#define GST_LOCAL_SINK(obj)           (G_TYPE_CHECK_INSTANCE_CAST((obj), GST_TYPE_LOCAL_SINK, GstLocalSink))
#define GST_LOCAL_SINK_CLASS(klass)   (G_TYPE_CHECK_CLASS_CAST((klass),  GST_TYPE_LOCAL_SINK, GstLocalSinkClass))
#define GST_IS_LOCAL_SINK(obj)        (G_TYPE_CHECK_INSTANCE_TYPE((obj), GST_TYPE_LOCAL_SINK))
#define GST_IS_LOCAL_SINK_CLASS(obj)  (G_TYPE_CHECK_CLASS_TYPE((klass),  GST_TYPE_LOCAL_SINK))
typedef struct _GstLocalSink GstLocalSink;
typedef struct _GstLocalSinkClass GstLocalSinkClass;

struct _GstLocalSink
{
  GstBaseSink base_localsink;

  GstLocalSurface *surface;
  char *channel;

  // local copy of the info-structs for read-only accesses
  // without locking the surface
  GstVideoInfo video_info;
  GstAudioInfo audio_info;
  guint type;
  guint queue_limit;
};

struct _GstLocalSinkClass
{
  GstBaseSinkClass base_localsink_class;
};

GType gst_local_sink_get_type (void);

G_END_DECLS
#endif


src/gstlocalsurface.c
/* GStreamer
 * Copyright (C) 2011 David A. Schleef <ds@schleef.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Suite 500,
 * Boston, MA 02110-1335, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <string.h>

#include "gstlocalsurface.h"

static GList *list;
static GMutex mutex;

GstLocalSurface *
gst_local_surface_get (const char *name)
{
  GList *g;
  GstLocalSurface *surface;

  g_mutex_lock (&mutex);
  for (g = list; g; g = g_list_next (g)) {
    surface = g->data;
    if (strcmp (name, surface->name) == 0) {
      surface->ref_count++;
      g_mutex_unlock (&mutex);
      return surface;
    }
  }

  surface = g_malloc0 (sizeof (GstLocalSurface));
  surface->ref_count = 1;
  surface->name = g_strdup (name);
  surface->queue_limit = 0;
  surface->queue =
      gst_data_queue_new (gst_local_surface_queue_test_isfull, NULL, NULL,
      surface);

  g_mutex_init (&surface->mutex);
  list = g_list_append (list, surface);
  g_mutex_unlock (&mutex);

  return surface;
}

gboolean
gst_local_surface_queue_test_isfull (GstDataQueue * queue,
    guint visible, guint bytes, guint64 time, gpointer checkdata)
{
  GstLocalSurface *surface = checkdata;
  return (surface->queue_limit != 0) && (visible >= surface->queue_limit);
}

guint
gst_local_surface_queue_level (GstLocalSurface * surface)
{
  GstDataQueueSize level;
  gst_data_queue_get_level (surface->queue, &level);
  return level.visible;
}

void
gst_local_surface_queue_push (GstLocalSurface * surface, GstBuffer * buffer)
{
  GstDataQueueItem *item = g_slice_new (GstDataQueueItem);

  item->destroy = (GDestroyNotify) gst_local_surface_audio_queue_free;

  gst_buffer_ref (buffer);
  item->object = GST_MINI_OBJECT (buffer);
  item->size = gst_buffer_get_size (buffer);
  item->duration = GST_BUFFER_DURATION (buffer);

  if (item->duration == GST_CLOCK_TIME_NONE) {
    item->duration = 0;
  }

  item->visible = TRUE;

  if (!gst_data_queue_push (surface->queue, item)) {
    item->destroy (item);
  }
}

GstBuffer *
gst_local_surface_queue_pop (GstLocalSurface * surface)
{
  GstDataQueueItem *item = NULL;

  if (gst_data_queue_is_empty (surface->queue))
    return NULL;

  if (!gst_data_queue_pop (surface->queue, &item))
    return NULL;

  if (item == NULL)
    return NULL;

  g_assert (GST_IS_BUFFER (item->object));
  return GST_BUFFER (item->object);
}

void
gst_local_surface_audio_queue_free (GstDataQueueItem * item)
{
  if (item->object) {
    gst_mini_object_unref (item->object);
  }

  g_slice_free (GstDataQueueItem, item);
}

void
gst_local_surface_unref (GstLocalSurface * surface)
{
  /* Mutex needed here, otherwise refcount might become 0
   * and someone else requests the same surface again before
   * we remove it from the list */
  g_mutex_lock (&mutex);
  if ((--surface->ref_count) == 0) {
    GList *g;

    for (g = list; g; g = g_list_next (g)) {
      GstLocalSurface *tmp = g->data;
      if (strcmp (tmp->name, surface->name) == 0) {
        list = g_list_delete_link (list, g);
        break;
      }
    }

    g_mutex_clear (&surface->mutex);
    gst_object_unref (surface->queue);
    g_free (surface->name);
    g_free (surface);
  }
  g_mutex_unlock (&mutex);
}


src/gstlocalsurface.h
/* GStreamer
 * Copyright (C) 2015 Peter Körner <peter@mazdermind.de>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifndef _GST_LOCAL_SURFACE_H_
#define _GST_LOCAL_SURFACE_H_

#include <gst/base/gstdataqueue.h>
#include <gst/audio/audio.h>
#include <gst/video/video.h>

#define GST_LOCAL_SURFACE_TYPE_UNDEFINED 0
#define GST_LOCAL_SURFACE_TYPE_VIDEO 1
#define GST_LOCAL_SURFACE_TYPE_AUDIO 2

// undefined -> waitforsrc/sink -> buffering -> running
#define GST_LOCAL_SURFACE_STATE_UNDEFINED 0
#define GST_LOCAL_SURFACE_STATE_WAITFORSRC 1
#define GST_LOCAL_SURFACE_STATE_WAITFORSINK 2
#define GST_LOCAL_SURFACE_STATE_BUFFERING 3
#define GST_LOCAL_SURFACE_STATE_RUNNING 4
#define GST_LOCAL_SURFACE_STATE_UNDERFLOW 5

G_BEGIN_DECLS typedef struct _GstLocalSurface GstLocalSurface;

struct _GstLocalSurface
{
  GMutex mutex;
  gint ref_count;

  char *name;
  guint state;
  guint type;
  guint queue_limit;

  GstVideoInfo video_info;
  GstAudioInfo audio_info;

  GstDataQueue *queue;
};

GstLocalSurface *gst_local_surface_get (const char *name);

void gst_local_surface_unref (GstLocalSurface * surface);

gboolean
gst_local_surface_queue_test_isfull (GstDataQueue * queue,
    guint visible, guint bytes, guint64 time, gpointer checkdata);

guint gst_local_surface_queue_level (GstLocalSurface * surface);

void gst_local_surface_audio_queue_free (GstDataQueueItem * item);

void
gst_local_surface_queue_push (GstLocalSurface * surface, GstBuffer * buffer);

GstBuffer *gst_local_surface_queue_pop (GstLocalSurface * surface);

G_END_DECLS
#endif


src/gstlocalsink.c
/* GStreamer
 * Copyright (C) 2015 Peter Körner <peter@mazdermind.de>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Suite 500,
 * Boston, MA 02110-1335, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <string.h>
#include "gstlocalsink.h"

GST_DEBUG_CATEGORY_STATIC (gst_local_sink_debug_category);
#define GST_CAT_DEFAULT gst_local_sink_debug_category


/* prototypes */
static void gst_local_sink_set_property (GObject * object,
    guint property_id, const GValue * value, GParamSpec * pspec);
static void gst_local_sink_get_property (GObject * object,
    guint property_id, GValue * value, GParamSpec * pspec);
static void gst_local_sink_finalize (GObject * object);

static void gst_local_sink_get_times (GstBaseSink * sink,
    GstBuffer * buffer, GstClockTime * start, GstClockTime * end);
static gboolean gst_local_sink_start (GstBaseSink * sink);
static gboolean gst_local_sink_stop (GstBaseSink * sink);
static gboolean gst_local_sink_set_caps (GstBaseSink * sink, GstCaps * caps);
static GstFlowReturn gst_local_sink_render (GstBaseSink * sink,
    GstBuffer * buffer);

enum
{
  PROP_0,
  PROP_CHANNEL,
  PROP_JITTERBUFFER
};

#define DEFAULT_CHANNEL ("default")
#define DEFAULT_JITTERBUFFER (50)
#define AVCAPS (GST_VIDEO_CAPS_MAKE (GST_VIDEO_FORMATS_ALL) ";" \
  GST_AUDIO_CAPS_MAKE (GST_AUDIO_FORMATS_ALL))


/* pad templates */
static GstStaticPadTemplate gst_local_sink_sink_template =
GST_STATIC_PAD_TEMPLATE ("sink",
    GST_PAD_SINK,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS (AVCAPS));

/* class initialization */
G_DEFINE_TYPE (GstLocalSink, gst_local_sink, GST_TYPE_BASE_SINK);


/* element boilerplate */
static void
gst_local_sink_class_init (GstLocalSinkClass * klass)
{
  GObjectClass *gobject_class = G_OBJECT_CLASS (klass);
  GstElementClass *element_class = GST_ELEMENT_CLASS (klass);
  GstBaseSinkClass *base_sink_class = GST_BASE_SINK_CLASS (klass);

  GST_DEBUG_CATEGORY_INIT (gst_local_sink_debug_category,
      "localsink", 0, "debug category for localsink element");

  gst_element_class_add_pad_template (element_class,
      gst_static_pad_template_get (&gst_local_sink_sink_template));

  gst_element_class_set_static_metadata (element_class,
      "inter-pipeline communication sink",
      "Sink",
      "virtual sink for timestamp aware inter-pipeline communication",
      "Peter Körner <peter@mazdermind.de>");

  gobject_class->set_property = gst_local_sink_set_property;
  gobject_class->get_property = gst_local_sink_get_property;
  gobject_class->finalize = gst_local_sink_finalize;
  base_sink_class->start = GST_DEBUG_FUNCPTR (gst_local_sink_start);
  base_sink_class->set_caps = GST_DEBUG_FUNCPTR (gst_local_sink_set_caps);
  base_sink_class->get_times = GST_DEBUG_FUNCPTR (gst_local_sink_get_times);
  base_sink_class->render = GST_DEBUG_FUNCPTR (gst_local_sink_render);
  base_sink_class->stop = GST_DEBUG_FUNCPTR (gst_local_sink_stop);

  g_object_class_install_property (gobject_class, PROP_CHANNEL,
      g_param_spec_string ("channel", "Channel",
          "Channel name to match local src and sink elements",
          DEFAULT_CHANNEL, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));

  g_object_class_install_property (gobject_class, PROP_JITTERBUFFER,
      g_param_spec_uint ("jitter-buffer", "Jitter-Buffer",
          "Number of Buffers to keep in the Jitter-Buffer between the sink and the source, 0 for infinity",
          0, G_MAXUINT, DEFAULT_JITTERBUFFER,
          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
}

static void
gst_local_sink_init (GstLocalSink * localsink)
{
  localsink->channel = g_strdup (DEFAULT_CHANNEL);
  GST_DEBUG_OBJECT (localsink, "init with default channel-name \"%s\"",
      localsink->channel);
  localsink->queue_limit = DEFAULT_JITTERBUFFER;
  GST_DEBUG_OBJECT (localsink, "init with default jitter-buffer-limit %u",
      localsink->queue_limit);
}

void
gst_local_sink_set_property (GObject * object, guint property_id,
    const GValue * value, GParamSpec * pspec)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (object);

  switch (property_id) {
    case PROP_CHANNEL:
      g_free (localsink->channel);
      localsink->channel = g_value_dup_string (value);
      GST_DEBUG_OBJECT (localsink, "setting channel-name to %s",
          localsink->channel);
      break;
    case PROP_JITTERBUFFER:
      localsink->queue_limit = g_value_get_uint (value);
      GST_DEBUG_OBJECT (localsink, "setting jitter-buffer-limit to %u",
          localsink->queue_limit);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
  }
}

void
gst_local_sink_get_property (GObject * object, guint property_id,
    GValue * value, GParamSpec * pspec)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (object);

  switch (property_id) {
    case PROP_CHANNEL:
      g_value_set_string (value, localsink->channel);
      break;
    case PROP_JITTERBUFFER:
      g_value_set_uint (value, localsink->queue_limit);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
  }
}

void
gst_local_sink_finalize (GObject * object)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (object);
  GST_DEBUG_OBJECT (localsink, "finalize");

  /* clean up object here */
  g_free (localsink->channel);

  G_OBJECT_CLASS (gst_local_sink_parent_class)->finalize (object);
}


/* basesink function implementations */
static gboolean
gst_local_sink_start (GstBaseSink * sink)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (sink);
  GST_DEBUG_OBJECT (localsink, "start");

  localsink->surface = gst_local_surface_get (localsink->channel);
  GST_DEBUG_OBJECT (localsink,
      "selected surface at %p for channel-name %s", localsink->surface,
      localsink->channel);

  g_mutex_lock (&localsink->surface->mutex);
  localsink->surface->queue_limit = localsink->queue_limit;
  if (localsink->surface->state == GST_LOCAL_SURFACE_STATE_UNDEFINED) {
    GST_DEBUG_OBJECT (localsink,
        "surface is in state UNDEFINED, changing state to WAITFORSRC");

    localsink->surface->state = GST_LOCAL_SURFACE_STATE_WAITFORSRC;
  } else if (localsink->surface->state == GST_LOCAL_SURFACE_STATE_WAITFORSINK) {
    GST_DEBUG_OBJECT (localsink,
        "surface is in state WAITFORSINK, changing state to BUFFERING");

    localsink->surface->state = GST_LOCAL_SURFACE_STATE_BUFFERING;
  } else {
    GST_ERROR_OBJECT (localsink,
        "surface is already connected, failing to connect another sink to it");
    g_mutex_unlock (&localsink->surface->mutex);
    return FALSE;
  }

  GST_DEBUG_OBJECT (localsink,
      "successfully connected to surface at %p for channel-name %s",
      localsink->surface, localsink->channel);

  g_mutex_unlock (&localsink->surface->mutex);
  return TRUE;
}

static gboolean
gst_local_sink_set_caps (GstBaseSink * sink, GstCaps * caps)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (sink);

  GstStructure *structure = gst_caps_get_structure (caps, 0);

  if (g_str_has_prefix (gst_structure_get_name (structure), "video/")) {
    GST_DEBUG_OBJECT (sink, "detected incoming caps as video");
    localsink->type = GST_LOCAL_SURFACE_TYPE_VIDEO;

    if (!gst_video_info_from_caps (&localsink->video_info, caps)) {
      GST_ERROR_OBJECT (sink,
          "failed to parse caps to video_info: %" GST_PTR_FORMAT, caps);
      return FALSE;
    }
  }

  else if (g_str_has_prefix (gst_structure_get_name (structure), "audio/")) {
    GST_DEBUG_OBJECT (sink, "detected incoming caps as audio");
    localsink->type = GST_LOCAL_SURFACE_TYPE_AUDIO;

    if (!gst_audio_info_from_caps (&localsink->audio_info, caps)) {
      GST_ERROR_OBJECT (sink,
          "failed to parse caps to audio_info: %" GST_PTR_FORMAT, caps);
      return FALSE;
    }
  }

  else {
    GST_ERROR_OBJECT (sink, "detected totally bogus caps: %" GST_PTR_FORMAT,
        caps);
    return FALSE;
  }

  g_mutex_lock (&localsink->surface->mutex);
  localsink->surface->video_info = localsink->video_info;
  localsink->surface->audio_info = localsink->audio_info;
  localsink->surface->type = localsink->type;
  g_mutex_unlock (&localsink->surface->mutex);

  return TRUE;
}

static void
gst_local_sink_get_times (GstBaseSink * sink, GstBuffer * buffer,
    GstClockTime * start, GstClockTime * end)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (sink);

  if (GST_BUFFER_TIMESTAMP_IS_VALID (buffer)) {
    *start = GST_BUFFER_TIMESTAMP (buffer);

    if (GST_BUFFER_DURATION_IS_VALID (buffer)) {
      *end = *start + GST_BUFFER_DURATION (buffer);
    }

    else if (localsink->type == GST_LOCAL_SURFACE_TYPE_VIDEO) {
      GST_WARNING_OBJECT (sink,
          "estimating buffer duration based on video-framerate");

      if (localsink->video_info.fps_n > 0) {
        *end = *start +
            gst_util_uint64_scale_int (GST_SECOND, localsink->video_info.fps_d,
            localsink->video_info.fps_n);
      }
    }

    else if (localsink->type == GST_LOCAL_SURFACE_TYPE_AUDIO) {
      GST_WARNING_OBJECT (sink,
          "estimating buffer duration based on audio-samplerate");

      if (localsink->audio_info.rate > 0) {
        *end = *start +
            gst_util_uint64_scale_int (gst_buffer_get_size (buffer), GST_SECOND,
            localsink->audio_info.rate * localsink->audio_info.bpf);
      }
    }
  }
}

static GstFlowReturn
gst_local_sink_render (GstBaseSink * sink, GstBuffer * buffer)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (sink);

  // maybe that lock is not needed and may harm the performance a lot
  g_mutex_lock (&localsink->surface->mutex);

  if (localsink->surface->state == GST_LOCAL_SURFACE_STATE_BUFFERING ||
      localsink->surface->state == GST_LOCAL_SURFACE_STATE_RUNNING) {

    GST_DEBUG_OBJECT (localsink,
        "pushing a buffer onto the jitter-buffer. level is currently at %u buffers, limit is %u",
        gst_local_surface_queue_level (localsink->surface),
        localsink->queue_limit);

    gst_local_surface_queue_push (localsink->surface, buffer);
  } else {

    GST_INFO_OBJECT (localsink,
        "surface it not connected to a source, throwing away buffer.");

  }

  g_mutex_unlock (&localsink->surface->mutex);

  return GST_FLOW_OK;
}

static gboolean
gst_local_sink_stop (GstBaseSink * sink)
{
  GstLocalSink *localsink = GST_LOCAL_SINK (sink);

  g_mutex_lock (&localsink->surface->mutex);

  if (localsink->surface->state == GST_LOCAL_SURFACE_STATE_BUFFERING) {
    GST_DEBUG_OBJECT (localsink,
        "surface is in state BUFFERING, changing state to WAITFORSINK");

    localsink->surface->state = GST_LOCAL_SURFACE_STATE_WAITFORSINK;
  } else if (localsink->surface->state == GST_LOCAL_SURFACE_STATE_WAITFORSRC) {
    GST_DEBUG_OBJECT (localsink,
        "surface is in state WAITFORSRC, changing state to UNDEFINED");

    localsink->surface->state = GST_LOCAL_SURFACE_STATE_UNDEFINED;
  }

  g_mutex_unlock (&localsink->surface->mutex);
  gst_local_surface_unref (localsink->surface);
  localsink->surface = NULL;

  return TRUE;
}
